Build error on Windows 10 dekstop

Hi, managed to run the build, but not sure why i am encountering this error
rcldotnet_common [ Exited with code 1 ]
towards the end of the build; and the rest of the rcldotnet_examples, rosidl_generator_dotnet all are not built.

`

Starting >>> rcl_yaml_param_parser
Failed <<< rcldotnet_common [ Exited with code 1 ]
Aborted <<< rosidl_generator_dds_idl
Aborted <<< rcpputils
Aborted <<< rcl_logging_spdlog
Aborted <<< rcl_yaml_param_parser
Aborted <<< rcl_logging_log4cxx
Aborted <<< rcl_logging_noop
Aborted <<< rosidl_generator_c
Summary: 87 packages finished [2min 10s]
1 package failed: rcldotnet_common
7 packages aborted: rcl_logging_log4cxx rcl_logging_noop rcl_logging_spdlog rcl_yaml_param_parser rcpputils rosidl_generator_c rosidl_generator_dds_idl
60 packages not processed
`

same for me.

```
Starting >>> rcl_logging_log4cxx
Starting >>> rcl_logging_noop
Starting >>> rcl_logging_spdlog
Starting >>> rcl_yaml_param_parser
--- stderr: rcldotnet_common
CMake Error at C:/dev/ros2_dotnet_ws/install/share/dotnet_cmake_module/cmake/Modules/FindDotNETExtra.cmake:16 (include):
  include called with wrong number of arguments.  include() only takes one
  file.
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)


CMake Error at C:/dev/ros2_dotnet_ws/install/share/dotnet_cmake_module/cmake/Modules/FindDotNETExtra.cmake:26 (csharp_add_project):
  Unknown CMake command "csharp_add_project".
Call Stack (most recent call first):
  CMakeLists.txt:17 (add_dotnet_library)


---
Failed   <<< rcldotnet_common   [ Exited with code 1 ]
Aborted  <<< rosidl_generator_dds_idl
Aborted  <<< rosidl_generator_c
Aborted  <<< rcpputils
Aborted  <<< rcl_logging_spdlog
Aborted  <<< rcl_yaml_param_parser
Aborted  <<< rcl_logging_log4cxx
Aborted  <<< rcl_logging_noop

Summary: 87 packages finished [1min 53s]
  1 package failed: rcldotnet_common
  7 packages aborted: rcl_logging_log4cxx rcl_logging_noop rcl_logging_spdlog rcl_yaml_param_parser rcpputils rosidl_generator_c rosidl_generator_dds_idl
  1 package had stderr output: rcldotnet_common
  60 packages not processed
```

At least the rcldotnet_common error was solved for the UWP build -- it was an incompatibility with the platform. (To get the specific error, add to the colcon command the option
`--event-handlers console_cohesion+`; in this way the errors printed in the standard output are also displayed.)

The next error now is on fastrtps and I suspect a difference with the language being used, given that it gives some syntax errors. 

e.g., error C2332: 'struct': missing tag name (compiling source file ros2-uwp\src\eProsima\Fast-RTPS\src\cpp\transport\TCPAcceptorSecure.cpp) 

I will keep digging and see what I will find.

@quattrinili 

Thanks for your report, though it'd better to file a separate ticket, given that the issue you encountered differs from this ticket.

> The next error now is on fastrtps and I suspect a difference with the language being used, given that it gives some syntax errors.
> 
> e.g., error C2332: 'struct': missing tag name (compiling source file ros2-uwp\src\eProsima\Fast-RTPS\src\cpp\transport\TCPAcceptorSecure.cpp)

FastRTPS keeps breaking the UWP support in every release, despite @theseankelly and I having sent patches to fix it, but they always revert the changes in new releases and don't even care about regression tests.

Just to make sure this is not a new issue in FastRTPS, can you check the following?

- Are you using our fork of FastRTPS? https://github.com/ros2-dotnet/Fast-RTPS

- Are you running this with the repos file for UWP? https://github.com/ros2-dotnet/ros2_dotnet/blob/master/ros2_dotnet_uwp.repos

Thanks!

--- a/.github/workflows/build_desktop.yml
+++ b/.github/workflows/build_desktop.yml
@@ -7,9 +7,6 @@ on:
 
 jobs:
   build-desktop:
-    strategy:
-      matrix:
-        arch: [x64]
     runs-on: windows-latest
     env:
       VisualStudioVersion: '16.0'
@@ -18,20 +15,26 @@ jobs:
     - name: Checkout source
       uses: actions/checkout@v2
 
-    - name: Install ROS2 dependencies
-      run: |
-        python -m pip install -U setuptools pip
-        pip install -U vcstool
-        pip install -U colcon-common-extensions
-        pip install -U catkin_pkg EmPy lark-parser pyparsing pyyaml numpy
-        pip install -U pytest coverage mock
-        pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8 pydocstyle
-        choco install -y cppcheck
+    - name: Setup ROS2
+      uses: ros-tooling/setup-ros@0.0.20
 
-    - name: Checkout ROS2 source
+    # Building ros2_dotnet as an overlay on top of the ROS2 binary distribution
+    # presents various challenges. Instead of using the
+    # 'with.required-ros-distributions' option, build a minimal set of packages
+    # (up to rcl, no testing packages) from source.
+    - name: Build ROS2
       run: |
         md \dev\ros2\src
         cd \dev\ros2
+        curl -sk https://raw.githubusercontent.com/ros2/ros2/eloquent/ros2.repos -o ros2.repos
+        vcs import src < ros2.repos
+        colcon build --event-handlers console_cohesion+ console_package_list+ --merge-install --packages-up-to rcl --cmake-args -DBUILD_TESTING=OFF -DTHIRDPARTY=ON -DINSTALL_EXAMPLES=OFF
+      shell: cmd
+
+    - name: Configure ros2_dotnet workspace
+      run: |
+        md \dev\ros2_dotnet_ws\src
+        cd \dev\ros2_dotnet_ws
         copy ${{github.workspace}}\ros2_dotnet.repos .
         vcs import src < ros2_dotnet.repos
         rd /s /q src\ros2_dotnet\ros2_dotnet
@@ -39,22 +42,16 @@ jobs:
         vcs branch
       shell: cmd
 
-    - name: Build ROS2 source (${{ matrix.arch }})
-      env:
-        TARGET_ARCH: ${{ matrix.arch }}
+    - name: Build ros2_dotnet source
       run: |
-        cd \dev\ros2
-        colcon build --event-handlers console_cohesion+ console_package_list+ --packages-ignore rmw_fastrtps_dynamic_cpp rcl_logging_log4cxx rcl_logging_spdlog ros2trace tracetools_launch tracetools_read tracetools_test tracetools_trace ros2lifecycle_test_fixtures --cmake-args -A %TARGET_ARCH% -DBUILD_TESTING=OFF -DTHIRDPARTY=ON -DINSTALL_EXAMPLES=OFF -DRCL_LOGGING_IMPLEMENTATION=rcl_logging_noop
+        call \dev\ros2\install\setup.bat
+        cd \dev\ros2_dotnet_ws
+        colcon build --event-handlers console_cohesion+ console_package_list+ --merge-install
       shell: cmd
 
-    - name: Compress ROS2 install directory
+    - name: Run tests
       run: |
-        Compress-Archive -Path \dev\ros2\install ros2.zip
-
-    - name: Upload artifacts
-      uses: actions/upload-artifact@v1
-      with:
-        name: ros2-desktop-${{ matrix.arch }}.zip
-        path: ros2.zip
-
-
+        call \dev\ros2\install\setup.bat
+        cd \dev\ros2_dotnet_ws
+        colcon test --event-handlers console_cohesion+ --return-code-on-test-failure --merge-install --packages-select rcldotnet
+      shell: cmd

--- a/.github/workflows/build_desktop.yml
+++ b/.github/workflows/build_desktop.yml
@@ -15,6 +15,10 @@ jobs:
     - name: Checkout source
       uses: actions/checkout@v2
 
+    - uses: actions/setup-python@v2
+      with:
+        python-version: '3.7.6'
+
     - name: Setup ROS2
       uses: ros-tooling/setup-ros@0.0.20
 

--- a/.github/workflows/build_linux.yml
+++ b/.github/workflows/build_linux.yml
@@ -1,0 +1,52 @@
+name: Build (Linux)
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build-linux:
+    runs-on: ubuntu-18.04
+
+    steps:
+    - name: Checkout source
+      uses: actions/checkout@v2
+
+    # https://docs.microsoft.com/en-us/dotnet/core/install/linux-package-manager-ubuntu-1804
+    - name: Install .NET SDK v2.1
+      run: |
+        wget https://packages.microsoft.com/config/ubuntu/18.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
+        sudo dpkg -i packages-microsoft-prod.deb
+        sudo add-apt-repository universe
+        sudo apt-get update
+        sudo apt-get install -y apt-transport-https
+        sudo apt-get update
+        sudo apt-get install -y dotnet-sdk-2.1
+
+    - name: Setup ROS2
+      uses: ros-tooling/setup-ros@0.0.20
+      with:
+        required-ros-distributions: eloquent
+
+    - name: Configure ros2_dotnet workspace
+      run: |
+        mkdir -p ~/ros2_dotnet_ws/src
+        cd ~/ros2_dotnet_ws
+        cp ${{github.workspace}}/ros2_dotnet.repos .
+        vcs import src < ros2_dotnet.repos
+        rm -rf src/ros2_dotnet/ros2_dotnet
+        cp -r ${{github.workspace}} src/ros2_dotnet/ros2_dotnet
+        vcs branch
+
+    - name: Build ros2_dotnet source
+      run: |
+        source /opt/ros/eloquent/setup.bash
+        cd ~/ros2_dotnet_ws
+        colcon build --event-handlers console_cohesion+ console_package_list+
+
+    - name: Run tests
+      run: |
+        source /opt/ros/eloquent/setup.bash
+        cd ~/ros2_dotnet_ws
+        colcon test --event-handlers console_cohesion+ --return-code-on-test-failure --packages-select rcldotnet

--- a/README.md
+++ b/README.md
@@ -8,7 +8,7 @@ Build status
 |----------|--------|
 | **Universal Windows Platform (x86/x64)** | ![Build (UWP)](https://github.com/ros2-dotnet/ros2_dotnet/workflows/Build%20(UWP)/badge.svg) |
 | **Windows Desktop (x64)** | ![Build (Desktop)](https://github.com/ros2-dotnet/ros2_dotnet/workflows/Build%20(Desktop)/badge.svg) |
-| **Linux** | (TODO) |
+| **Linux** | ![Build (Desktop)](https://github.com/ros2-dotnet/ros2_dotnet/workflows/Build%20(Linux)/badge.svg) |
 
 Introduction
 ------------
@@ -41,38 +41,72 @@ Lots of things!
 Sounds great, how can I try this out?
 -------------------------------------
 
-First of all install the standard ROS2 dependencies for your operating system of choice https://github.com/ros2/ros2/wiki/Installation#building-from-source
+First of all install the standard ROS2 dependencies for your operating system
+of choice (**NOTE**: only do this if building Windows Desktop or Linux. For UWP,
+see the relevant section below): https://github.com/ros2/ros2/wiki/Installation#building-from-source
 
-Next make sure you've either installed .Net Core (preferred) https://www.microsoft.com/net/learn/get-started or Mono https://www.mono-project.com/
+Next make sure you've either installed .Net Core (preferred)
+https://www.microsoft.com/net/learn/get-started or Mono
+https://www.mono-project.com/. (**NOTE**: For building unit tests, .NET 2.1 is
+required).
 
-The following steps show how to build the examples on Windows and Linux:
+For running on Linux or Windows Desktop, one can build `ros2_dotnet` (along with
+all desired packages containing interface definitions) as an overlay on top
+of an existing ROS2 installation. The `ros2_dotnet.repos` contains all
+necessary repositories to build the core `ros2_dotnet` project along with all
+standard ROS2 interface packages. If you are using other packages which provide
+interface definitions, those must also be included in the `ros2_dotnet` workspace
+in order for .NET bindings to be generated. (NOTE: if you wish to build the
+core of ROS2 from source, everything through the `rcl` layer is required.)
+
+For running within UWP (Universal Windows Platform) applications, the entire
+core of ROS2 must be compiled for UWP compatibility.
 
 Windows (Desktop)
 -----------------
+Assuming you've installed ROS2 (pre-built binary packages) to the directory
+c:\dev\ros2_eloquent per the official [installation instructions](https://index.ros.org/doc/ros2/Installation/Eloquent/Windows-Install-Binary/),
+run the following from an Administrator Visual Studio 2019 Developer Command
+Prompt:
+
+(**NOTE**: Building as an overlay on top of *binary* distributions of ROS2 has
+presented some challenges. As of this writing, you may also need to include the
+`rosidl` package due to some unicode/locale compatibility problems.
+This is done for you below in the line preceding `colcon build`. This step
+can/should be omitted if building on top of a built-from-source ROS2 workspace)
 
 ```
+call \dev\ros2_eloquent\local_setup.bat
 md \dev\ros2_dotnet_ws\src
 cd \dev\ros2_dotnet_ws
-curl -sk https://raw.githubusercontent.com/esteve/ros2_dotnet/master/ros2_dotnet.repos -o ros2_dotnet.repos
+curl -sk https://raw.githubusercontent.com/ros2-dotnet/ros2_dotnet/master/ros2_dotnet.repos -o ros2_dotnet.repos
 vcs import \dev\ros2_dotnet_ws\src < ros2_dotnet.repos
+git clone --branch eloquent https://github.com/ros2/rosidl src\ros2\rosidl
 colcon build --merge-install
 ```
+
 
 Linux
 -----
-
+Assuming ROS2 eloquent installed to the standard location, run the following commands:
 ```
+source /opt/ros/eloquent/setup.bash
 mkdir -p ~/ros2_dotnet_ws/src
 cd ~/ros2_dotnet_ws
-wget https://raw.githubusercontent.com/esteve/ros2_dotnet/master/ros2_dotnet.repos
+wget https://raw.githubusercontent.com/ros2-dotnet/ros2_dotnet/master/ros2_dotnet.repos
 vcs import ~/ros2_dotnet_ws/src < ros2_dotnet.repos
-colcon build --merge-install
+colcon build
 ```
 
-Universal Windows Platform (ARM, Win32, Win64)
+Universal Windows Platform (Win32, Win64)
 ----------------------------------------------
+We'll build this in two steps, first `ament` (the build system) and related tools
+which will run natively on the host, followed by ROS2 itself, built for UWP and
+the target architecture.
 
-We'll build this in two steps, first `ament` (the build system) and then ROS2 itself. This is because `ament` will run on the host platform, whereas `ros2-dotnet` on the target.
+If you have previously installed ROS2 dependencies (OpenSSL, tinyxml, log4cxx,
+etc) it is strongly recommended to uninstall those dependencies before building
+to avoid any non-UWP binaries getting pulled into the build.
 
 ament
 -----
@@ -80,7 +114,7 @@ ament
 ```
 md \dev\ament\src
 cd \dev\ament
-curl -sk https://raw.githubusercontent.com/esteve/ros2_dotnet/master/ament_dotnet_uwp.repos -o ament_dotnet_uwp.repos
+curl -sk https://raw.githubusercontent.com/ros2-dotnet/ros2_dotnet/master/ament_dotnet_uwp.repos -o ament_dotnet_uwp.repos
 vcs import src < ament_dotnet_uwp.repos
 colcon build --merge-install
 call install\local_setup.bat
@@ -94,7 +128,7 @@ Replace `%TARGET_ARCH%` with Win32 or x64
 ```
 md \dev\ros2\src
 cd \dev\ros2
-curl -sk https://raw.githubusercontent.com/esteve/ros2_dotnet/master/ros2_dotnet_uwp.repos -o ros2_dotnet_uwp.repos
+curl -sk https://raw.githubusercontent.com/ros2-dotnet/ros2_dotnet/master/ros2_dotnet_uwp.repos -o ros2_dotnet_uwp.repos
 vcs import src < ros2_dotnet_uwp.repos
 cd \dev\ament
 call install\local_setup.bat

--- a/ros2_dotnet.repos
+++ b/ros2_dotnet.repos
@@ -1,151 +1,23 @@
 repositories:
-  ament/ament_cmake:
-    type: git
-    url: https://github.com/ament/ament_cmake.git
-    version: eloquent
-  ament/ament_index:
-    type: git
-    url: https://github.com/ament/ament_index.git
-    version: eloquent
-  ament/ament_lint:
-    type: git
-    url: https://github.com/ament/ament_lint.git
-    version: eloquent
-  ament/ament_package:
-    type: git
-    url: https://github.com/ament/ament_package.git
-    version: eloquent
-  ament/googletest:
-    type: git
-    url: https://github.com/ament/googletest.git
-    version: ros2
-  ament/osrf_pycommon:
-    type: git
-    url: https://github.com/osrf/osrf_pycommon.git
-    version: eloquent
-  ament/uncrustify_vendor:
-    type: git
-    url: https://github.com/ament/uncrustify_vendor.git
-    version: eloquent
   ament_dotnet/ament_cmake_export_assemblies:
     type: git
     url: https://github.com/esteve/ament_cmake_export_assemblies.git
     version: master
-  eProsima/Fast-CDR:
-    type: git
-    url: https://github.com/eProsima/Fast-CDR.git
-    version: v1.0.11
-  eProsima/Fast-RTPS:
-    type: git
-    url: https://github.com/ros2-dotnet/Fast-RTPS.git
-    version: eloquent-windows
-  eProsima/foonathan_memory_vendor:
-    type: git
-    url: https://github.com/eProsima/foonathan_memory_vendor.git
-    version: master
-  micro-ROS/ros_tracing/ros2_tracing:
-    type: git
-    url: https://gitlab.com/micro-ROS/ros_tracing/ros2_tracing.git
-    version: 0.2.10
-  ros2/ament_cmake_ros:
-    type: git
-    url: https://github.com/ros2/ament_cmake_ros.git
-    version: eloquent
-  ros2/rcpputils:
-    type: git
-    url: https://github.com/ros2/rcpputils.git
-    version: eloquent
-  ros2/rcutils:
-    type: git
-    url: https://github.com/ros2/rcutils.git
-    version: eloquent
   ros2/common_interfaces:
     type: git
     url: https://github.com/ros2/common_interfaces.git
-    version: eloquent
-  ros2/example_interfaces:
-    type: git
-    url: https://github.com/ros2/example_interfaces.git
-    version: eloquent
-  ros2/launch:
-    type: git
-    url: https://github.com/ros2/launch.git
-    version: eloquent
-  ros2/libyaml_vendor:
-    type: git
-    url: https://github.com/ros2/libyaml_vendor.git
-    version: eloquent
-  ros2/poco_vendor:
-    type: git
-    url: https://github.com/ros2/poco_vendor.git
-    version: eloquent
-  ros2/python_cmake_module:
-    type: git
-    url: https://github.com/ros2/python_cmake_module.git
-    version: eloquent
-  ros2/rcl:
-    type: git
-    url: https://github.com/ros2/rcl.git
     version: eloquent
   ros2/rcl_interfaces:
     type: git
     url: https://github.com/ros2/rcl_interfaces.git
     version: eloquent
-  ros2/rclpy:
-    type: git
-    url: https://github.com/ros2/rclpy.git
-    version: eloquent
-  ros2/rcl_logging:
-    type: git
-    url: https://github.com/ros2/rcl_logging.git
-    version: eloquent
-  ros2/rmw:
-    type: git
-    url: https://github.com/ros2/rmw.git
-    version: eloquent
-  ros2/rmw_fastrtps:
-    type: git
-    url: https://github.com/ros2/rmw_fastrtps.git
-    version: eloquent
-  ros2/rmw_implementation:
-    type: git
-    url: https://github.com/ros2/rmw_implementation.git
-    version: eloquent
-  ros2/ros2cli:
-    type: git
-    url: https://github.com/ros2/ros2cli.git
-    version: eloquent
-  ros2/rosidl:
-    type: git
-    url: https://github.com/ros2/rosidl.git
-    version: eloquent
-  ros2/rosidl_dds:
-    type: git
-    url: https://github.com/ros2/rosidl_dds.git
-    version: eloquent
   ros2/rosidl_defaults:
     type: git
     url: https://github.com/ros2/rosidl_defaults.git
     version: eloquent
-  ros2/rosidl_python:
-    type: git
-    url: https://github.com/ros2/rosidl_python.git
-    version: eloquent
-  ros2/rosidl_typesupport:
-    type: git
-    url: https://github.com/ros2-dotnet/rosidl_typesupport.git
-    version: eloquent-windows
-  ros2/rosidl_typesupport_fastrtps:
-    type: git
-    url: https://github.com/ros2/rosidl_typesupport_fastrtps.git
-    version: eloquent
   ros2/test_interface_files:
     type: git
     url: https://github.com/ros2/test_interface_files.git
-    version: eloquent
-  ros2/tinydir_vendor:
-    type: git
-    url: https://github.com/ros2/tinydir_vendor.git
     version: eloquent
   ros2/unique_identifier_msgs:
     type: git

